README.md
# FOREWORD #
Based on https://github.com/h3imdall/ftp-remote-edit, I added the 'folder' concept:

 * you can manage folders from Packages -> Ftp-Remote-Plus -> edit folders
 * for each server you can select one folder from the dropdown list
 * folders are displayed in white into tree view in alphabetic order
 * orphan servers (not associated to any folder) are displayed after folders
 * change of folder name is propagated to associated servers
 * deleting folder makes children servers orphan

 TODO:

 * autoRevealActiveFile is disabled because it messes up at tree load time

 Below is the original README content of the package I cloned and hacked

# Ftp-Remote-Edit-Plus

Editing files on your server without the need for creating a local project. It is not necassary to download all files of your project.
Simply connect and edit your remote files. The files will be automatically updated to the server on saving.

![A screenshot of your package](https://raw.githubusercontent.com/h3imdall/ftp-remote-edit/master/screenshot.png)

## Whats new in 0.12

- Added Option "Open in Atom Dock" - Open the view as tab in atom dock instead of panel. Only available from Atom 1.17.0
- Added FTP Message Log - Shows sent/received FTP commands in log
- Changed refreshing behavior of the tree view
- Feature request: Drag-and-Drop (issue-25)
- Feature request: Copy / Cut / Paste files (issue-24)
- Feature request: Rename copied file if exists to prevent collision of names (issue-61)
- Feature request: Statusbar for upload/download (issue-69)
- Feature request: Option "Disable Notification on save/upload" (issue-55)
- Feature request: Option "Hide Ignored Names" (issue-78)
- Feature request: Option "Opening hidden files" (e.g. htaccess) (issue-66)
- Feature request: Option "Preview opened file in pending pane" to have the same behavior than in atom (issue-83)
- Feature request: Option "Sort Servers By Name" (issue-106)
- Feature request: Fuzzy finder for remote files (issue-109)
- Feature request: Support for file-icons package (issue-125)
- Feature request: Option "Auto Reveal Active File" (issue-158)
- Feature request: Open from command line (issue-165)
- Feature request: Option "Toggle on atom startup" (issue-168)
- Feature request: "Configuration view - Duplicate Server" (issue-77)
- Feature request: "Change permissions of a file" (issue-137)

## Getting started

- Toggle the view with "ftp-remote-edit:toggle" or use keybinding `ctrl-space`
- Enter the master password. If not allready set, enter the firsttime. All information about your server settings will be encrypted with this password.
- Right click and select "Edit Servers" to open the configuration view. Here you can add, edit and delete your (s)ftp server settings.

## Keybindings

- Toggle the view with `ctrl-space`
- Toggle the fuzzy finder with `ctrl-alt-p` (item must be selected in the tree view)

## Helpfull commands
- Toggle the view with "ftp-remote-edit:toggle"
- Change master password with "ftp-remote-edit:change-password"
- Toggle the fuzzy finder with "ftp-remote-edit:finder"
- Reindex the fuzzy finder cache with "ftp-remote-edit:finder-reindex-cache"

## URI handler
Add temporary server for ftp/sftp by using uri. It is possible to use it with/without username, password and port.
- atom://ftp-remote-edit/sftp://username:password@host:port
- atom://ftp-remote-edit/ftp://username:password@host:port

## Package preferences

- `Tree View` - `Open On Startup` - Open the view automatically when atom starts.
- `Tree View` - `Open In Atom Dock` - Open the view as tab in atom dock instead of panel. Only available from Atom 1.17.0
- `Tree View` - `Show On Right Side` - Show the view on the right side of the editor instead of the left.
- `Tree View` - `Allow Pending Pane Items` - Allow items to be previewed without adding them to a pane permanently.
- `Tree View` - `Hide Ignored Files` - Don't show items matched by the `Ignored Names` core config setting.
- `Tree View` - `Show Hidden Files` - Force FTP Server to show hidden files (e.g. htaccess)
- `Tree View` - `Sort Folders Before Files` - When listing directory items, list subdirectories before listing files.
- `Tree View` - `Sort Servers By Name` - When listing servers items, list servers by name rather than by host.
- `Tree View` - `Auto Reveal Active File` - Auto reveal the current active file on the tree view.
- `Tree View Finder` - `Key For Search` - Specifies the key at which the search is to be used.
- `Tree View Finder` - `Ignored Names` - Files and directories matching these patterns and the `Ignored Names` core config setting will be ignored during indexing.
- `Notification` - `Successful Upload` - Show notification on successful upload
- `Notification` - `Failed Upload` - Open protocol view in case of failed upload.
- `Development` - `Debug Mode` - Output debug messages to the console.

## I'd like to support this project

Help us bring this project to the moon! Atom's rocket needs to get somewhere, right?

- **Contribute!** I'll be happy to accept pull requests!
- **Bug hunting!** [Report](https://github.com/h3imdall/ftp-remote-edit/issues) them!
- **Feature request?** [Please let me know](https://github.com/h3imdall/ftp-remote-edit/issues) by filling an issue!
- **Star this project** on [Atom](https://atom.io/packages/ftp-remote-edit), [Github](https://github.com/h3imdall/ftp-remote-edit)
- **Donate for this project** - [![Donate](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=KAMKHBBJH7KB2)

## Special Thanks

- [@miles-collier](https://github.com/miles-collier) [Issue-38](https://github.com/h3imdall/ftp-remote-edit/pull/38) Keybinding for toggling
- [@dustinparker](https://github.com/dustinparker) [Issue-106](https://github.com/h3imdall/ftp-remote-edit/pull/112) Sort servers by name
- [@Me1onRind](https://github.com/Me1onRind)
[Issue-109](https://github.com/h3imdall/ftp-remote-edit/pull/124) Fuzzy finder for remote files
- [@wacki4](https://github.com/wacki4)
[Issue-165](https://github.com/h3imdall/ftp-remote-edit/pull/169) Open from command line
- [@FabrizioCaldarelli ](https://github.com/FabrizioCaldarelli)
[Issue-173](https://github.com/h3imdall/ftp-remote-edit/pull/178),  [Issue-44](https://github.com/h3imdall/ftp-remote-edit/pull/174) Help fix some errors


CHANGELOG.md
## 1.2.0 - Release
- Added sync icon into saving file tab

## 1.1.3 - Release
- Bug fix on opening servers view with no folders saved

## 1.1.0 - Release
- First Release


lib/views/tree-view.js
'use babel';

import FolderView from './folder-view.js';
import ServerView from './server-view.js';
import DirectoryView from './directory-view.js';
import FileView from './file-view.js';
import FtpLogView from './ftp-log-view.js';
import { $, ScrollView } from 'atom-space-pen-views';
import { basename, dirname, trailingslashit, cleanJsonString } from './../helper/format.js';
import { decrypt, encrypt, checkPassword, setPassword, b64EncodeUnicode, b64DecodeUnicode } from './../helper/secure.js';
import { resetIgnoredPatterns, resetIgnoredFinderPatterns } from './../helper/helper.js';
import { throwErrorIssue44 } from './../helper/issue.js';

const Path = require('path');
const resizeCursor = process.platform === 'win32' ? 'ew-resize' : 'col-resize';
const FTP_REMOTE_EDIT_URI = 'h3imdall://ftp-remote-edit';

class TreeView extends ScrollView {

  TreeView() {
    super.TreeView();

    const self = this;
    self.folders = null;
    self.servers = null;
  };

  static content() {
    return this.div({
      class: 'ftp-remote-edit-view ftp-remote-edit-resizer tool-panel',
      'data-show-on-right-side': atom.config.get('ftp-remote-edit-plus.tree.showOnRightSide'),
    }, () => {
      this.div({
        class: 'ftp-remote-edit-scroller order--center',
        outlet: 'scroller',
      }, () => {
        this.ol({
          class: 'ftp-remote-edit-list full-menu list-tree has-collapsable-children focusable-panel',
          tabindex: -1,
          outlet: 'list',
        });
      });

      this.div({
        class: 'ftp-remote-edit-resize-handle',
        outlet: 'horizontalResize',
      });

      this.subview('ftpLogView', new FtpLogView());

      this.div({
        class: 'info',
        tabindex: -1,
        outlet: 'info',
      });
    });
  };

  getTitle() {
    return "Ftp-Remote-Edit";
  }

  getURI() {
    return FTP_REMOTE_EDIT_URI;
  }

  getAllowedLocations() {
    return ["left", "right"];
  }

  getDefaultLocation() {
    if (atom.config.get('ftp-remote-edit-plus.tree.showOnRightSide')) {
      return "right";
    } else {
      return "left";
    }
  }

  isPermanentDockItem() {
    return true;
  }

  initialize(state) {
    super.initialize(state)

    const self = this;

    let html = '<background-tip>';
    html += '<ul class="centered background-message">';
    html += '<li class="message fade-in">You can edit the servers from the Settings View with ftp-remote-edit:edit-servers<br/><br/><a role="configure" class="btn btn-xs btn-default icon">Edit Servers</a></li>';
    html += '</ul>';
    html += '</background-tip>';
    self.info.html(html);

    if (!self.servers) {
      self.showInfo();
    } else {
      self.hideInfo();
    }

    // Events
    atom.config.onDidChange('ftp-remote-edit.tree.showOnRightSide', () => {
      self.element.dataset.showOnRightSide = atom.config.get('ftp-remote-edit-plus.tree.showOnRightSide');
      if (self.isVisible() && atom.config.get('ftp-remote-edit-plus.tree.showInDock') == false) {
        self.detach();
        self.attach();
      }
    });
    atom.config.onDidChange('ftp-remote-edit.tree.sortFoldersBeforeFiles', () => {
      if (self.isVisible()) {
        self.reload();
      }
    });
    atom.config.onDidChange('ftp-remote-edit.tree.sortServerProfilesByName', () => {
      if (self.isVisible()) {
        self.reload();
      }
    });
    atom.config.onDidChange('ftp-remote-edit.tree.hideIgnoredNames', () => {
      resetIgnoredPatterns();
      if (self.isVisible()) {
        self.reload();
      }
    });
    atom.config.onDidChange('core.ignoredNames', () => {
      resetIgnoredPatterns();
      resetIgnoredFinderPatterns();
      if (self.isVisible()) {
        self.reload();
      }
    });
    atom.config.onDidChange('ftp-remote-edit.finder.ignoredNames', () => {
      resetIgnoredPatterns();
      resetIgnoredFinderPatterns();
      if (self.isVisible()) {
        self.reload();
      }
    });

    self.on('mousedown', (e) => {
      var entryToSelect;
      if (entryToSelect = e.target.closest('.entry')) {
        e.stopPropagation();
        this.selectEntry(entryToSelect);
      }
    });

    // Info Panel
    self.info.on('click', '[role="configure"]', (e) => {
      atom.commands.dispatch(atom.views.getView(atom.workspace), 'ftp-remote-edit:edit-servers');
    });
    self.info.on('click', '[role="toggle"]', (e) => {
      self.toggle();
    });

    // Resize Panel
    self.horizontalResize.on('dblclick', (e) => {
      self.resizeToFitContent(e);
    });
    self.horizontalResize.on('mousedown', (e) => {
      self.resizeHorizontalStarted(e);
    });

    // Keyboard Navigation
    self.list.on('keydown', (e) => { self.remoteKeyboardNavigation(e); });
  };

  attach() {
    const self = this;
    if (atom.config.get('ftp-remote-edit-plus.tree.showOnRightSide')) {
      self.panel = atom.workspace.addRightPanel({
        item: self
      });
    } else {
      self.panel = atom.workspace.addLeftPanel({
        item: self
      });
    }
  };

  detach() {
    const self = this;

    if (self.panel) {
      self.panel.destroy();
      self.panel = null;
    }
  };

  destroy() {
    const self = this;

    if (self.list) {
      if (self.list.children()) {
        self.list.children()
          .detach();
      }
      self.servers = null;
    }
    self.remove();
  };

  toggle() {
    const self = this;
    if (atom.config.get('ftp-remote-edit-plus.tree.showInDock')) {
      atom.workspace.toggle(this);
    } else {
      if (self.isVisible()) {
        self.detach();
      } else {
        self.attach();
        self.resizeToFitContent();
      }
    }
  };

  loadFolders(password) {
    const self = this;

    // Add detail information in error notification
    // Uncaught SyntaxError: Unexpected token  in JSON at position 0 #44
    // https://github.com/h3imdall/ftp-remote-edit/issues/44
    let configHash = atom.config.get('ftp-remote-edit-plus.folders');
    if (configHash) {
      let config = decrypt(password, configHash);
      try {
        self.folders = JSON.parse(cleanJsonString(config));
        self.folders.sort((a, b) => {
          if (a.name.toLowerCase() < b.name.toLowerCase()) return -1;
          if (a.name.toLowerCase() > b.name.toLowerCase()) return 1;
          return 0;
        });
        //load servers
        self.loadServers(password);
        if(typeof self.servers != 'undefined') {
          //attach servers associated to a folder to the proper folder
          self.folders.forEach((config,index) => {
            self.folders[index].servers = self.servers[config.id];
          });
          //set folder NOT associated to any folder to be displayed
          self.servers = self.servers[null];
        }
        return true;
      } catch (e) {
        throwErrorIssue44(e, password);
        return false;
      }
    }
    return true;
  }

  loadServers(password) {
    const self = this;

    // Add detail information in error notification
    // Uncaught SyntaxError: Unexpected token  in JSON at position 0 #44
    // https://github.com/h3imdall/ftp-remote-edit/issues/44
    let configHash = atom.config.get('ftp-remote-edit-plus.servers');
    if (configHash) {
      let config = decrypt(password, configHash);
      let sortServerProfilesByName = atom.config.get('ftp-remote-edit-plus.tree.sortServerProfilesByName');

      try {
        servers = JSON.parse(cleanJsonString(config));
        self.servers = {};
        servers.forEach((config,index) => {
          //self.folders[index].servers = self.servers;
          if(!self.servers[config.folder]) {
              self.servers[config.folder] = [];
          }
          self.servers[config.folder].push(config);
        });
        /*self.servers.sort((a, b) => {
          if (sortServerProfilesByName) {
            if (a.name.toLowerCase() < b.name.toLowerCase()) return -1;
            if (a.name.toLowerCase() > b.name.toLowerCase()) return 1;
          } else {
            if (a.host < b.host) return -1;
            if (a.host > b.host) return 1;
          }
          return 0;
      });*/
        return true;

      } catch (e) {
        throwErrorIssue44(e, password);
        return false;
      }
    }
    return true;
  }

  reload() {
    const self = this;
    if (!self.folders) {
      self.list.children().detach();
      self.showInfo();
      return;
    }

    self.list.children().detach();
    self.hideInfo();
    if(self.folders && self.folders.length > 0) {
        self.folders.forEach((config) => {
          self.addFolder(config);
        });
    }
    if(self.servers && self.servers.length > 0) {
        self.servers.forEach((config) => {
          self.addServer(config);
        });
    }
  };

  addFolder(config) {
    const self = this;

    let folder = new FolderView(config, self);
    self.list.append(folder);
  };

  addServer(config) {
    const self = this;

    let server = new ServerView(config, self);
    self.list.append(server);
  };

  showInfo() {
    this.info.css('display', 'flex');
  };

  hideInfo() {
    this.info.hide();
  };

  deselect(elementsToDeselect) {
    let i, len, selected;
    if (elementsToDeselect == null) {
      elementsToDeselect = $('.ftp-remote-edit-view .selected');
    }
    for (i = 0, len = elementsToDeselect.length; i < len; i++) {
      selected = elementsToDeselect[i];
      selected.classList.remove('selected');
    }
    return void 0;
  };

  selectEntry(entry) {
    if (entry == null) {
      return;
    }
    this.deselect();
    try {
      if (entry.classList) {
        entry.classList.add('selected');
      } else {
        entry.addClass('selected');
      }
    } catch (e) {}
    return entry;
  }

  getSelected() {
    let selected;
    selected = $('.ftp-remote-edit-view .selected');
    if (selected) {
      return selected[0];
    }
    return null;
  }

  getServerByLocalPath(pathOnFileSystem) {
    const self = this;

    return new Promise((resolve, reject) => {

      if (!self.servers) reject();
      if (!self.list) reject();

      self.servers.forEach((config) => {
        let server = new ServerView(config, self);
        let path = server.getLocalPath(true);
        if (pathOnFileSystem.indexOf(path) != -1) {
          resolve(server);
        }
      });
      reject();
    });
  }

  getElementByLocalPath(pathOnFileSystem, root, type = 'directory') {
    const self = this;

    let elementparent = null;
    let elementname = basename(pathOnFileSystem, Path.sep);
    let elementpath = dirname(pathOnFileSystem, Path.sep) + elementname;
    let dirpath = dirname(pathOnFileSystem, Path.sep);

    let a = trailingslashit(pathOnFileSystem, Path.sep);
    let b = trailingslashit(root.getLocalPath(true), Path.sep);
    if (a == b) {
      return self.getServerView(root.config, root.treeView);
    } else if (type == 'file') {
      return self.getFileView(self.getElementByLocalPath(dirpath, root), {
        name: elementname,
        path: elementpath,
        size: 0,
        rights: null
      });
    } else {
      return self.getDirectoryView(self.getElementByLocalPath(dirpath, root), {
        name: elementname,
        path: elementpath,
        rights: null
      });
    }
  }

  getServerView(config, self) {
    return new ServerView(config, self);
  }

  getDirectoryView(parent, directory) {
    return new DirectoryView(parent, directory);
  }

  getFileView(parent, file) {
    return new FileView(parent, file);
  }

  resizeHorizontalStarted(e) {
    e.preventDefault();

    this.resizeWidthStart = this.width();
    this.resizeMouseStart = e.pageX;
    $(document)
      .on('mousemove', this.resizeHorizontalView.bind(this));
    $(document)
      .on('mouseup', this.resizeHorizontalStopped);
  };

  resizeHorizontalStopped() {
    delete this.resizeWidthStart;
    delete this.resizeMouseStart;
    $(document)
      .off('mousemove', this.resizeHorizontalView);
    $(document)
      .off('mouseup', this.resizeHorizontalStopped);
  };

  resizeHorizontalView(e) {
    if (e.which !== 1) {
      return this.resizeHorizontalStopped();
    }

    let delta = e.pageX - this.resizeMouseStart;
    let width = 0;
    if (atom.config.get('ftp-remote-edit-plus.tree.showOnRightSide')) {
      width = Math.max(50, this.resizeWidthStart - delta);
    } else {
      width = Math.max(50, this.resizeWidthStart + delta);
    }

    this.width(width);
  };

  resizeToFitContent(e) {
    if (e) e.preventDefault();

    if (atom.config.get('ftp-remote-edit-plus.tree.showInDock')) {
      const paneContainer = atom.workspace.paneContainerForItem(this)
      // NOTE: This is an internal API access
      // It's necessary because there's no Public API for it yet
      if (paneContainer && typeof paneContainer.state.size === 'number' && paneContainer.widthOrHeight == 'width' && typeof paneContainer.render === 'function') {
        paneContainer.state.size = 1
        paneContainer.state.size = (this.list.outerWidth() + 10);
        paneContainer.render(paneContainer.state)
      }
    } else {
      this.width(1);
      this.width(this.list.outerWidth() + 10);
    }
  };

  remoteKeyboardNavigation(e) {
    let arrows = { left: 37, up: 38, right: 39, down: 40 },
      keyCode = e.keyCode || e.which;

    switch (keyCode) {
      case arrows.up:
        this.remoteKeyboardNavigationUp();
        break;
      case arrows.down:
        this.remoteKeyboardNavigationDown();
        break;
      case arrows.left:
        this.remoteKeyboardNavigationLeft();
        break;
      case arrows.right:
        this.remoteKeyboardNavigationRight();
        break;
      default:
        return;
    }

    e.preventDefault();
    e.stopPropagation();
    this.remoteKeyboardNavigationMovePage();
  };

  remoteKeyboardNavigationUp() {
    let current = this.list.find('.selected'),
      next = current.prev('.entry:visible');
    if (next.length) {
      while (next.is('.expanded') && next.find('.entries .entry:visible').length) {
        next = next.find('.entries .entry:visible');
      }
    } else {
      next = current.closest('.entries').closest('.entry:visible');
    }
    if (next.length) {
      current.removeClass('selected');
      next.last().addClass('selected');
    }
  };

  remoteKeyboardNavigationDown() {
    let current = this.list.find('.selected'),
      next = current.find('.entries .entry:visible');
    if (!next.length) {
      tmp = current;

      // Workaorund skip after 10
      let counter = 1;
      do {
        next = tmp.next('.entry:visible');
        if (!next.length) {
          tmp = tmp.closest('.entries').closest('.entry:visible');
        }
        counter++;
      } while (!next.length && !tmp.is('.project-root') && counter < 10);
    }
    if (next.length) {
      current.removeClass('selected');
      next.first().addClass('selected');
    }
  };

  remoteKeyboardNavigationLeft() {
    const current = this.list.find('.selected');
    if (current.is('.file')) {
      parent = current.view().parent.view();
      parent.collapse();
      current.removeClass('selected');
      parent.addClass('selected');
    } else if (current.is('.directory') && current.view().isExpanded) {
      current.view().collapse();
    } else if (current.is('.directory') && !current.view().isExpanded) {
      parent = current.view().parent.view();
      parent.collapse();
      current.removeClass('selected');
      parent.addClass('selected');
    } else {
      current.view().collapse();
    }
  };

  remoteKeyboardNavigationRight() {
    const current = this.list.find('.selected');
    if (current.is('.directory') || current.is('.server')) {
      if (!current.view().isExpanded) {
        current.view().expand();
      }
    } else {
      current.view().open();
    }
  };

  remoteKeyboardNavigationMovePage() {
    const current = this.list.find('.selected');
    if (current.length) {
      let scrollerTop = this.scroller.scrollTop(),
        selectedTop = current.position().top;
      if (selectedTop < scrollerTop - 10) {
        this.scroller.pageUp();
      } else if (selectedTop > scrollerTop + this.scroller.height() - 10) {
        this.scroller.pageDown();
      }
    }
  };
}

module.exports = TreeView;
